services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,238 @@
+
+    @Autowired
+    private MessageProducer messageProducer;
+
+    /**
+     * Creates storage for chains of messages.
+     */
+    @PostConstruct
+    public void setUp() {
+        messagesChains = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new HashMap<>());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<InfoData> sendAndGet(String topic, Message message) {
+        CompletableFuture<InfoData> future = new CompletableFuture<>();
+
+        ListenableFuture<SendResult<String, Message>> futureResult = messageProducer.send(topic, message);
+        futureResult.addCallback(
+                sentResult -> pendingRequests.put(message.getCorrelationId(), future),
+                error -> future.completeExceptionally(new MessageNotSentException(error.getMessage()))
+        );
+
+        return future.whenComplete((response, error) -> pendingRequests.remove(message.getCorrelationId()));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<List<InfoData>> sendAndGetChunked(String topic, Message message) {
+        CompletableFuture<List<InfoData>> future = new CompletableFuture<>();
+
+        ListenableFuture<SendResult<String, Message>> futureResult = messageProducer.send(topic, message);
+        futureResult.addCallback(
+                sentResult -> {
+                    pendingChunkedRequests.put(message.getCorrelationId(), future);
+                    messagesChains.put(message.getCorrelationId(), new ArrayList<>());
+                },
+                error -> future.completeExceptionally(new MessageNotSentException(error.getMessage()))
+        );
+
+        return future.whenComplete((response, error) -> pendingChunkedRequests.remove(message.getCorrelationId()));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void send(String topic, Message message) {
+        messageProducer.send(topic, message);
+    }
+
+    /**
+     * Processes messages that come back to NB topics, usually messages come as a response to some request.
+     * If this response is for pended request then such request will be marked as completed.
+     * Messages might be chunked and not chunked. If chunked we need to wait until we receive the last one
+     * and only then collect all responses and complete the request.
+     *
+     * @param message received message.
+     */
+    void onResponse(Message message) {
+        if (!isValid(message)) {
+            logger.warn("Skipping invalid message: {}", message);
+            return;
+        }
+
+        if (message instanceof ErrorMessage) {
+            ErrorMessage error = (ErrorMessage) message;
+            logger.error("Response message is error: {}", error);
+
+            completeWithError(error);
+        } else if (message instanceof InfoMessage) {
+            if (pendingChunkedRequests.containsKey(message.getCorrelationId())) {
```

Comment:
It will fail it someone send 2 message with same correlation id via `sendAndGet` and via `sendAndGetChunked` at same time.